File: tests/test_app_state.py
"""Tests for application state and persistence."""

from pathlib import Path

from src.state.app_state import AppState


def test_initial_state():
    state = AppState()
    assert state.current_folder is None
    assert state.files == []
    assert state.index == 0
    assert state.view_mode == "grid"
    assert state.zoom == 1.0


def test_navigate_next(sample_images, tmp_image_dir):
    state = AppState()
    state.load_folder(tmp_image_dir)
    assert state.index == 0
    state.navigate_next()
    assert state.index == 1
    state.navigate_next()
    assert state.index == 2


def test_navigate_prev(sample_images, tmp_image_dir):
    state = AppState()
    state.load_folder(tmp_image_dir)
    state.index = 5
    state.navigate_prev()
    assert state.index == 4


def test_navigate_next_clamps(sample_images, tmp_image_dir):
    state = AppState()
    state.load_folder(tmp_image_dir)
    state.index = len(state.files) - 1
    state.navigate_next()
    assert state.index == len(state.files) - 1


def test_navigate_prev_clamps(sample_images, tmp_image_dir):
    state = AppState()
    state.load_folder(tmp_image_dir)
    state.navigate_prev()
    assert state.index == 0


def test_current_file(sample_images, tmp_image_dir):
    state = AppState()
    state.load_folder(tmp_image_dir)
    assert state.current_file is not None
    assert state.current_file.suffix == ".jpg"


def test_load_folder_with_target(sample_images, tmp_image_dir):
    state = AppState()
    target = sample_images[4]
    state.load_folder(tmp_image_dir, target_file=target)
    assert state.index == 4


def test_persist_and_restore(tmp_path, sample_images, tmp_image_dir):
    config = tmp_path / "config"
    state = AppState(config_dir=config)
    state.load_folder(tmp_image_dir)
    state.save()

    state2 = AppState(config_dir=config)
    state2.restore()
    assert state2.last_folder == tmp_image_dir


def test_remove_file(sample_images, tmp_image_dir):
    state = AppState()
    state.load_folder(tmp_image_dir)
    count_before = len(state.files)
    removed = state.files[3]
    state.remove_file(removed)
    assert len(state.files) == count_before - 1
    assert removed not in state.files

File: tests/test_app_state.rs
use std::path::PathBuf;
use image::{DynamicImage, RgbImage};
use omarchy_imageview::state::app_state::{AppState, ViewMode};

fn save_img(dir: &std::path::Path, name: &str) -> PathBuf {
    let p = dir.join(name);
    DynamicImage::ImageRgb8(RgbImage::new(10, 10)).save(&p).unwrap();
    p
}

fn make_images(dir: &std::path::Path, count: usize) -> Vec<PathBuf> {
    (1..=count).map(|i| save_img(dir, &format!("photo{i}.jpg"))).collect()
}

#[test]
fn test_initial_state() {
    let state = AppState::new(Some(PathBuf::from("/tmp/test-state")));
    assert!(state.current_folder.is_none());
    assert!(state.files.is_empty());
    assert_eq!(state.index, 0);
    assert_eq!(state.view_mode, ViewMode::Grid);
    assert!(state.zoom_fit);
    assert_eq!(state.zoom, 1.0);
}

#[test]
fn test_navigate_next() {
    let dir = tempfile::tempdir().unwrap();
    make_images(dir.path(), 10);
    let mut state = AppState::new(Some(PathBuf::from("/tmp/test-nav")));
    state.load_folder(dir.path(), None);
    assert_eq!(state.index, 0);
    state.navigate_next();
    assert_eq!(state.index, 1);
    state.navigate_next();
    assert_eq!(state.index, 2);
}

#[test]
fn test_navigate_prev() {
    let dir = tempfile::tempdir().unwrap();
    make_images(dir.path(), 10);
    let mut state = AppState::new(Some(PathBuf::from("/tmp/test-prev")));
    state.load_folder(dir.path(), None);
    state.index = 5;
    state.navigate_prev();
    assert_eq!(state.index, 4);
}

#[test]
fn test_navigate_clamps() {
    let dir = tempfile::tempdir().unwrap();
    make_images(dir.path(), 5);
    let mut state = AppState::new(Some(PathBuf::from("/tmp/test-clamp")));
    state.load_folder(dir.path(), None);
    state.index = 4;
    state.navigate_next();
    assert_eq!(state.index, 4); // stays at end
    state.index = 0;
    state.navigate_prev();
    assert_eq!(state.index, 0); // stays at start
}

#[test]
fn test_current_file() {
    let dir = tempfile::tempdir().unwrap();
    make_images(dir.path(), 5);
    let mut state = AppState::new(Some(PathBuf::from("/tmp/test-cf")));
    state.load_folder(dir.path(), None);
    assert!(state.current_file().is_some());
}

#[test]
fn test_load_folder_with_target() {
    let dir = tempfile::tempdir().unwrap();
    let images = make_images(dir.path(), 10);
    let mut state = AppState::new(Some(PathBuf::from("/tmp/test-target")));
    state.load_folder(dir.path(), Some(&images[4]));
    assert_eq!(state.index, 4);
}

#[test]
fn test_persist_and_restore() {
    let dir = tempfile::tempdir().unwrap();
    make_images(dir.path(), 5);
    let config = tempfile::tempdir().unwrap();
    let mut state = AppState::new(Some(config.path().to_path_buf()));
    state.load_folder(dir.path(), None);
    state.save();

    let mut state2 = AppState::new(Some(config.path().to_path_buf()));
    state2.restore();
    assert_eq!(state2.last_folder, Some(dir.path().to_path_buf()));
}

#[test]
fn test_remove_file() {
    let dir = tempfile::tempdir().unwrap();
    make_images(dir.path(), 10);
    let mut state = AppState::new(Some(PathBuf::from("/tmp/test-rm")));
    state.load_folder(dir.path(), None);
    let count = state.files.len();
    let removed = state.files[3].clone();
    state.remove_file(&removed);
    assert_eq!(state.files.len(), count - 1);
    assert!(!state.files.contains(&removed));
}
